homeassistant/components/modbus/__init__.py

```diff
@@ -189,47 +194,59 @@ async def _connect_delay(self):
             await asyncio.sleep(self._config_delay)
             self._config_delay = 0
 
-    def setup(self):
-        """Set up pymodbus client."""
+    def _framer(self, method):
+        if method == "ascii":
+            return ModbusAsciiFramer(ClientDecoder())
+        elif method == "rtu":
+            return ModbusRtuFramer(ClientDecoder())
+        elif method == "binary":
+            return ModbusBinaryFramer(ClientDecoder())
+        elif method == "socket":
+            return ModbusSocketFramer(ClientDecoder())
+        else:
+            return None
+
+    async def create_serial(self):
+        """Create serial modbus client and connect."""
         # pylint: disable = E0633
         # Client* do deliver loop, client as result but
         # pylint does not accept that fact
 
+    async def setup(self, hass):
+        """Set up pymodbus client."""
         if self._config_type == "serial":
-            _, self._client = ClientSerial(
-                schedulers.ASYNC_IO,
-                method=self._config_method,
-                port=self._config_port,
+            # reconnect ??
+            framer = self._framer(self._config_method)
```

Comment:
How do you plan to set the framer for TCP/RTU/UDP? I see you're setting the framer only for serial connection.

Reply:
actually RTU is TCP with a different framer. The reason I do not set it, is that the pymodbus async code, do not set the framer.

This is of course another problem, and there are already a reported issue.

Reply:
OK, because now it's the whole static method `_framer` redundant. However, we can keep it, and use it when `pymodbus` fixes the issue.

Reply:
Not really redundant, the code is needed for serial :-) and instead of copying it into the serial “if” I prefer to have it as a function so it can be used in the others as well